File: temp/mapreduce.py
def map_function(employee):
    department = employee['department']
    age = employee['age']
    return (department, age)


def reduce_function(key, values):
    total_age = sum(values)
    count = len(values)
    average_age = total_age / count
    return (key, average_age)


# Input data (a list of dictionaries representing the JSON data)
data = [
    {"id": 1, "name": "Alice", "age": 30, "department": "HR"},
    {"id": 2, "name": "Bob", "age": 40, "department": "Finance"},
    {"id": 3, "name": "Charlie", "age": 35, "department": "IT"}
]

# Map Step
map_output = map(map_function, data)


# Organize the map output into a dictionary where the keys are departments
# and the values are lists of ages
reduce_input = {}
for department, age in map_output:
    if department not in reduce_input:
        reduce_input[department] = []
    reduce_input[department].append(age)

# # Reduce Step
# reduce_output = {key: reduce_function(key, values)
#                  for key, values in reduce_input.items()}

# # Print the output
# for department, average_age in reduce_output.items():
#     print(f"The average age in the {department} department is {average_age}")

# Modify reduce_function to return a dictionary


def reduce_function(key, values):
    total_age = sum(values)
    count = len(values)
    average_age = total_age / count
    return {"key": key, "averageAge": average_age}


# Reduce Step
reduce_output = [reduce_function(key, values)
                 for key, values in reduce_input.items()]
print(f"reduce_input{reduce_input}")
print(f"reduce_output{reduce_output}")

File: temp/mapreduce.js
function mapFunction(employee) {
  return { department: employee.department, age: employee.age };
}

function reduceFunction(key, values) {
  const totalAge = values.reduce((acc, val) => acc + val, 0);
  const count = values.length;
  const averageAge = totalAge / count;
  return { key, averageAge };
}

// Input data (an array of objects representing the JSON data)
const data = [
  { id: 1, name: "Alice", age: 30, department: "HR" },
  { id: 2, name: "Bob", age: 40, department: "Finance" },
  { id: 3, name: "Charlie", age: 35, department: "IT" },
];

// Map Step
const mapOutput = data.map(mapFunction);
console.log("map output", mapOutput);

// Organize the map output into an object where the keys are departments
// and the values are lists of ages
const reduceInput = {};
mapOutput.forEach(({ department, age }) => {
  if (!reduceInput[department]) {
    reduceInput[department] = [];
  }
  reduceInput[department].push(age);
});
console.log("reduce input", JSON.stringify(reduceInput, null, 2));

// Reduce Step
const reduceOutput = Object.entries(reduceInput).map(([key, values]) => {
  return reduceFunction(key, values); //[{"IT", 35}, {"Finance", 40}, {"HR", 30}]
});
console.log("reduceOutput: ", JSON.stringify(reduceOutput, null, 2));
// Print the output
reduceOutput.forEach(({ key, averageAge }) => {
  console.log(`The average age in the ${key} department is ${averageAge}`);
});
